src/utils/file.cpp

```diff
@@ -233,27 +276,139 @@ namespace file_util {
+      return "/";
+    }
+
+    return path.substr(0, pos);
+  }
+
   /**
    * Path expansion
    */
-  const string expand(const string& path) {
-    string ret;
-    vector<string> p_exploded = string_util::split(path, '/');
-    for (auto& section : p_exploded) {
-      switch(section[0]) {
-        case '$':
-          section = env_util::get(section.substr(1));
-          break;
-        case '~':
-          section = env_util::get("HOME");
-          break;
+  string expand(const string& path) {
+    if (path.empty()) {
+      return "/";
+    }
+
+    auto ret = path;
+    if (ret[0] == '~') {
+      file_util::expand_home_dir(ret);
+    }
+
+    vector<string> p_exploded = string_util::split(ret, '/');
+    for (auto it = p_exploded.begin(); it != p_exploded.end(); ) {
+      if (it->empty() || *it == ".") {
+        it = p_exploded.erase(it);
+      } else if (*it == "..") {
+        if (it != p_exploded.begin()) {
+          --it;
+          it = p_exploded.erase(it);
+        }
+
+        it = p_exploded.erase(it);
+      } else {
+        auto& path = *it;
+
+        // subtract 1 since we are not interested when the last character is '$'
+        for (auto pos = path.find('$'); pos < path.length() - 1; pos = path.find('$', pos + 1)) {
+          auto replace_start = pos++;
+
+          string::size_type replace_end, count;
+          if (path[pos] == '{') {
+            ++pos;
+            count = path.find('}', pos + 1);
+
+            if (count == string::npos) {
+              break;
+            }
+
+            count -= pos;
+            replace_end = pos + (count + 1);
+          } else {
+            count = path.length() - pos;
+            replace_end = pos + count;
+          }
+
+          auto env = env_util::get(path.substr(pos, count));
+          path.replace(replace_start, replace_end, env);
+
+          pos += count;
+        }
+
+        // erase any separators at the end (due to expansion) to avoid double '//'
+        while (!path.empty() && path.back() == '/') {
+            path.pop_back();
+        }
+
+        // remove separators at the start to avoid double '//'
+        string::size_type start = 0;
+        while (start < path.length() && path[start] == '/') {
+            ++start;
+        }
+
+        if (start > 0) {
+            path = path.substr(start);
+        }
+
+        ++it;
       }
     }
+
+    if (p_exploded.empty()) {
+        return "/";
+    }
+
     ret = string_util::join(p_exploded, "/");
-    if (ret[0] != '/')
+
+    // if the original path had a trailing separator, preserve it
+    if (path.back() == '/') {
+      ret += '/';
+    }
+
+    // if the path is not absolute, insert a separator at the start
+    if (ret[0] != '/') {
       ret.insert(0, 1, '/');
+    }
+
     return ret;
   }
+
+  /**
+   * Path expansion, making relative paths absolute
+   */
+  string expand_relative_to(const string& path, const string& base_dir) {
```

Comment:
Please also add tests here. I think there may be some issues if path is something like `../bin` or starts with an env variable.

Reply:
Good call with regards to the `../bin` scenario, that didn't work correctly. If the base_dir was something like `/usr/bin` and the relative path was `../lib`, the result wouldn't be `/usr/lib` as the it should have been. I refactored the `is_absolute` check to work on paths with unexpanded variables and then switched up the relative expansion logic so that it handles the above case correctly. Also added some test cases for the scenarios.